specs/interop/predeploys.md

```diff
@@ -254,6 +288,77 @@ Note that the `relayMessage` function is `payable` to enable relayers to earn in
+to the source chain, it's crucial to ensure the message can only be sent back
+to the `L2ToL2CrossDomainMessenger` contract in its source chain.
+
+This function has no auth, which allows anyone to send back a given message hash.
+The `RETURN_DELAY` variable is added to give the users enough time to replay their
+failed messages and to prevent malicious actors from performing a griefing attack
+by sending messages back immediately.
+
+Once the message is returned to the source chain, the message on the local chain is set
+as successful in the `successfulMessages` mapping to ensure non-replayability.
+
+```solidity
+function revert(bytes32 _messageHash) external nonReentrant {
+    if (successfulMessages[messageHash]) revert MessageAlreadyRelayed();
+
+    (uint256 messageTimestamp, uint256 messageSource) = failedMessages[messageHash];
+
+    // Return delay is necessary to avoid griefing attacks by providing users with time to
+    // replay their transactions give this function has no auth
+    if (block.timestamp <  messageTimestamp + RETURN_DELAY) revert DelayHasNotEnsued();
+
+    delete failedMessages[messageHash];
+
+    // This is set to true to ensure non-replayability on this chain
+    successfulMessages[messageHash] = true;
+
+    bytes memory data = abi.encodeCall(
+        L2ToL2CrossDomainMessenger.receiveMessageHash,
+        (messageSource, block.chainid, Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER, messageHash)
+    );
+    emit SentMessage(data);
+}
+```
+
+#### Storing Sent-back Message Hashes
+
+When receiving a sent-back message only message hashes
+of actual failed messages should be stored, for this we must ensure the origin,
+sender and caller are all the expected contracts.
+
+It's also important to ensure only the hashes of messages that were initiated
+in this chain are accepted and that they come from the correct destination
+chain id.
+
+If all checks have been successful, the message has is stored in the
+`returnedMessages` mapping. This enables smart contracts to read from it and
+check whether a message failed or not, and handle the failing case accordingly.
+
+```solidity
+function receiveMessageHash(uint256 _messageSource, uint256 _messageDestination, address _sender, bytes32 _messageHash) external {
+    if (_messageSource != block.chainid) revert IncorrectMessageSource();
+    if (_sender != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert SenderIsNotCrossDomainMesenger();
+    if (msg.sender != Predeploys.CROSS_L2_INBOX) revert ReceiveMessageHashCallerNotCrossL2Inbox();
+
+    if (CrossL2Inbox(Predeploys.CROSS_L2_INBOX).origin() != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+        revert CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+    }
+
+    if (CrossL2Inbox(Predeploys.CROSS_L2_INBOX).chainid() != _messageDestination) {
+        revert CrossL2InboxDestinationDoesntMatch();
+    }
+
+    returnedMessages[_messageHash] = block.timestamp;
+
+    emit MessageHashReceived(_messageHash);
```

Comment:
Can we update this event name to match the `revert` or `rollback` naming scheme?